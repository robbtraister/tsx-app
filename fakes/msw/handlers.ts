import { createHandler, Resource } from "./utils";

export const handlers = [
  ...createHandler(
    "/api/v1/pages",
    new Resource(() => import("../data/api/v1/pages.json"))
  ),
];
