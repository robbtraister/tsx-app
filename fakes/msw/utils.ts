import { rest } from "msw";

export function createHandler(
  endpoint: string,
  resource: {
    get: (id: string) => Promise<unknown>;
    list: (search: URLSearchParams) => Promise<unknown>;
  }
) {
  return [
    rest.get(endpoint, async (req, res, ctx) => {
      return res(ctx.json(await resource.list(req.url.searchParams)));
    }),
    rest.get(`${endpoint}/:id`, async (req, res, ctx) => {
      return res(ctx.json(await resource.get(req.params.id)));
    }),
  ];
}

export class Resource<Type extends { id: string } = { id: string }> {
  constructor(private load: () => Promise<{ default: Type[] }>) {}

  async get(id: string): Promise<Type | undefined> {
    const { default: data } = await this.load();
    return data.find((datum) => datum.id === id);
  }

  async list(_search: URLSearchParams): Promise<Type[]> {
    const { default: data } = await this.load();
    return data;
  }
}
