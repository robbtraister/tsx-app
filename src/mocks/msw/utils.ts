import { rest, RequestHandler } from "msw";

import { resources, Resource } from "../resources";

function createHandler(endpoint: string, resource: Resource) {
  return [
    rest.get(endpoint, async (req, res, ctx) => {
      const query = [...req.url.searchParams.entries()].reduce(
        (acc, [key, value]) => ({ ...acc, [key]: value }),
        {}
      );
      return res(ctx.json(await resource.list(query)));
    }),

    rest.get(`${endpoint}/:id`, async (req, res, ctx) => {
      return res(ctx.json(await resource.get(req.params.id)));
    }),
  ];
}

export const handlers = ([] as RequestHandler[]).concat(
  ...Object.entries(resources).map(([endpoint, resource]) =>
    createHandler(endpoint, resource)
  )
);
