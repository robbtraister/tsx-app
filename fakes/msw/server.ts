import { setupServer } from "msw/node";

import { handlers } from "./utils";

export const server = setupServer(...handlers);
