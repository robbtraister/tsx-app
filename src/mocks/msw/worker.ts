import { setupWorker } from "msw";

import { handlers } from "./utils";

export const worker = setupWorker(...handlers);
