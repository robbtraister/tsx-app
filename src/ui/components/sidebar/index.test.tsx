import { render, waitFor } from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";

import { StaticRouter } from "react-router";

import { Sidebar } from ".";

const server = setupServer(
  rest.get("/api/v1/pages", (_req, res, ctx) => {
    return res(ctx.json({ pages: ["one", "two"] }));
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe("sidebar component", () => {
  it("mocked", async () => {
    const { container } = render(
      <StaticRouter location="/">
        <Sidebar />
      </StaticRouter>
    );

    expect(container).toMatchSnapshot();

    await waitFor(() =>
      expect(container.querySelectorAll("li")).toHaveLength(2)
    );

    expect(container).toMatchSnapshot();
  });
});
