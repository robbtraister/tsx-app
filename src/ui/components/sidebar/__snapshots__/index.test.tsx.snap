// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sidebar component mocked 1`] = `
<div>
  <div
    class="sidebar"
    data-testid="sidebar"
  >
    <ul />
  </div>
</div>
`;

exports[`sidebar component mocked 2`] = `
<div>
  <div
    class="sidebar"
    data-testid="sidebar"
  >
    <ul>
      <li>
        <a
          href="/one"
        >
          one
        </a>
      </li>
      <li>
        <a
          href="/two"
        >
          two
        </a>
      </li>
    </ul>
  </div>
</div>
`;
